### PR TITLE
M znalezniak/gating network

### PR DESCRIPTION
**Pull request by XYZ321COD**

### CODE REVIEW

File: run_experiments.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 # call(["python", "run.py", "-dataset-name", "cifar10", "--b","128", "--epoch", "50", "--number_of_rsacm", "1", "--distill_type", "per_rsacm"])
 # call(["python", "run.py", "-dataset-name", "cifar10", "--b","128", "--epoch", "50", "--number_of_rsacm", "8", "--distill_type", "per_accm"])
 # call(["python", "run.py", "-dataset-name", "cifar10", "--b", "256", "--epoch", "35", "--number_of_rsacm", "8", "--distill_type", "per_rsacm", '--arch', 'resnet50'])
-call(["python", "run.py", "-dataset-name", "cifar10", "--b", "256", "--epoch", "35", "--number_of_rsacm", "8", "--distill_type", "per_rsacm", '--arch', 'resnet50', '--inference', '--load_student'])
+call(["python", "run.py", "-dataset-name", "cifar10", "--b", "16", "--epoch", "35", "--number_of_rsacm", "8", "--distill_type", "per_rsacm", '--arch', 'resnet50', '--load_student'])
```
Comment (bartwojcik) [inline run_experiments.py:10]: Why is our batch size (I am guessing it is BS) so small?

Comment (MichalZnalezniak) [reply run_experiments.py:10]: After adding gating networks I had to drastically lower the batch_size - on batch_size 256 getting out-of-memory errors.

File: run.py [hunk 2]
```diff
@@ -106,6 +108,7 @@ def main():
     #  It’s a no-op if the 'gpu_index' argument is a negative integer or None.
     with torch.cuda.device(args.gpu_index):
         enn = ENN(model=student_model, optimizer=optimizer, args=args)
+        enn.train_gating_networks(train_loader, valid_loader)
```
Comment (bartwojcik) [inline run.py:111]: No flag, and calling this before enn.train() ?

Comment (MichalZnalezniak) [reply run.py:111]: I will add the flag in the future. In general code of the repo requires cleaning up. The main reason for this PR is just to showcase the differences/code for the training of gating networks. I won't merge it before clean up.

File: models/gating_network.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import torch
+
+
+class Gating_Network(nn.Module):
+    def __init__(self, activation, args):
+        super().__init__()
+        m = nn.Conv2d(activation.shape[1], 30, 1, stride=2).cuda()
+        k = m(activation)
+        self.neurons_fc = k.shape[1] * k.shape[2] * k.shape[3]
+        self.first_conv = nn.Conv2d(activation.shape[1], 30, 1, stride=2).cuda()
```
Comment (bartwojcik) [inline models/gating_network.py:12]: Alternatively it is quite common to do global average (or max) pooling after a convolution that has tens (or even more) of output channels. Then as a bonus you don't have to do any inference (or calculation) of output size in the constructor, and it works for any input image size. On the other hand this is a different architecture that is not guaranteed to work better than this one, so I would leave it as it is for now.

File: models/gating_network.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        self.fc = nn.Linear(self.neurons_fc, args.number_of_rsacm).cuda()
+
+    def forward(self, x):
+        x = self.first_conv(x)
+        x = torch.flatten(x, 1)
```
Comment (bartwojcik) [inline models/gating_network.py:17]: Perhaps add an activation/relu here after first_conv.

File: models/accm.py [hunk 5]
```diff
@@ -30,14 +31,24 @@ def forward(self, x):
...
         self.h_channels = h_channels
         self.number_of_rsacm = number_of_rsacm
+        self.args = args
         self.rsacm_list = nn.ModuleList(
             [(RsacmBlock(input_layer, self.h_channels)) for i in range(0, self.number_of_rsacm)])
+        self.rsacm_cost = rsacm_cost
+
+    def initialize_gating_network(self, activation, args):
+        self.gating_network =  Gating_Network(activation, args)
+        self.optimizer_for_gating_network = torch.optim.Adam(self.gating_network.parameters(), args.lr, weight_decay=0)
```
Comment (bartwojcik) [inline models/accm.py:45]: Looks like this is redundant and unused now?

File: models/accm.py [hunk 6]
```diff
@@ -30,14 +31,24 @@ def forward(self, x):
...
 
     def forward(self, x):
+        if self.args.train_gating_networks:
+            gating_network_output =  self.gating_network(x)
+            gating_network_output_index = torch.argmax(gating_network_output, dim=1)
+            costs = torch.tensor([self.rsacm_cost*i for i in range(1,self.number_of_rsacm+1)]).cuda()
         x = x.detach() # Detach x make sure that gradient for this layer won't flow to previous layers.
```
Comment (bartwojcik) [inline models/accm.py:52]: We are training our gating network in a end-to-end manner, so this detach must not be called for gating network training.

File: models/accm.py (hunk 5 above)
Comment (MichalZnalezniak) [reply models/accm.py:45]: So 
`self.gating_network` is used but `self.optimizer_for_gating_network` is not used. I will remove `self.optimizer_for_gating_network` when doing clean-up.

File: models/accm.py [hunk 7]
```diff
@@ -30,14 +31,24 @@ def forward(self, x):
...
+        self.rsacm_cost = rsacm_cost
+
+    def initialize_gating_network(self, activation, args):
+        self.gating_network =  Gating_Network(activation, args)
+        self.optimizer_for_gating_network = torch.optim.Adam(self.gating_network.parameters(), args.lr, weight_decay=0)
 
     def forward(self, x):
+        if self.args.train_gating_networks:
+            gating_network_output =  self.gating_network(x)
+            gating_network_output_index = torch.argmax(gating_network_output, dim=1)
```
Comment (bartwojcik) [inline models/accm.py:50]: Argmax is not differentiable, and we want the gradient to flow to the gating network (to all of them, actually). Softmax is actually the differentiable approximation to argmax we want (and, in particular, gumbel-softmax). We can:
1. optionally leave the argmax for inference (so that we are honest in the fact that our model really saves FLOPs),
2. use the gating network output as-is (the output after gumbel-softmax).
I think that point 2 means that the gating network should return $N$ outputs for each example, instead of an index.

File: models/accm.py [hunk 8]
```diff
@@ -47,6 +58,14 @@ def forward(self, x):
             rsacm_output = self.rsacm_list[i](x_copy)
             self.x_list.append(rsacm_output + x.detach()) # Detach x before adding it to the sum 
             x = rsacm_output + x
+        self.x_list_tensor = torch.stack((self.x_list))
+        if self.args.train_gating_networks:
+            lists_of_outputs = []
+            for index2, elements in enumerate(gating_network_output_index):
+                lists_of_outputs.append(self.x_list_tensor.permute(1,0,2,3,4)[index2][elements])
```
Comment (bartwojcik) [inline models/accm.py:65]: We should multiply (select) the x_list_tensor by the gating_network_outputs (one-hot if hard=True) vectors instead of using gating_network_output_index. Can you print out the shapes of both these? They should be helpful in writing this line of code.

File: enn.py [hunk 9]
```diff
@@ -33,41 +36,112 @@ def __init__(self, *args, **kwargs):
 
 
 
+    def inference_nas(self, train_loader, valid_loader):
```
Comment (bartwojcik) [inline enn.py:39]: Seems that train_loader is unused in this function?

File: enn.py [hunk 10]
```diff
@@ -33,41 +36,112 @@ def __init__(self, *args, **kwargs):
...
+        for idx, value in enumerate(self.model.replaced_modules):
+            for number_of_rsacm in range(1,self.args.number_of_rsacm+1):
+                with torch.no_grad():
+                    get_module_by_name(self.model, value).override_forward(number_of_rsacm)
```
Comment (bartwojcik) [inline enn.py:43]: I am not sure that I understand this correctly function - this line sets the number of used rsacms only for the current moddule ("value" variable, from the outer loop)? So we choose the number of rsacms from 1 to 8 sequentially? Shouldn't we reset the number of rsacms (to one or max) for all other modules (other than the current one) in each iteration?

File: enn.py [hunk 11]
```diff
@@ -33,41 +36,112 @@ def __init__(self, *args, **kwargs):
...
+            losses_per_layers_student_train = {module_name: [[] for i in range(0,self.args.number_of_rsacm)] for module_name in self.model.replaced_modules}
+            for i, (images, labels) in enumerate(tqdm(train_loader)):
+                    images, labels = images.to(self.args.device), labels.to(self.args.device)
+                    self.teacher_model(images)
+                    loss_distill = 0
+                    for index, module_name in enumerate(self.model.replaced_modules):
+                        get_module_by_name(self.model, module_name)(self.teacher_model.inputs[index][0])
+                        for rsacm_index, rsacm_block_output in enumerate(get_module_by_name(self.model, module_name).x_list):
+                            loss_distill += self.criterion_distilation(rsacm_block_output, self.teacher_model.activation[index])
+                            losses_per_layers_student_train[module_name][rsacm_index].append(self.criterion_distilation(rsacm_block_output, self.teacher_model.activation[index]))
```
Comment (bartwojcik) [inline enn.py:76]: Minor, but duplicated code similar to lines 53-65, perhaps extract it to a function and run it once for train_loader and second time for valid_loader?

File: enn.py [hunk 12]
```diff
@@ -33,41 +36,112 @@ def __init__(self, *args, **kwargs):
...
+                plt.plot()
+                plt.xlabel('Number of RSACM')
+                plt.ylabel('MSELoss value')
+                plt.savefig(f'./graphs/{module_name}.png')
+                plt.clf()
+
+
+    def train_gating_networks(self, train_loader, valid_loader):
+
+        accuracy_student = eval(valid_loader, self.model, self.args)
```
Comment (bartwojcik) [inline enn.py:104]: This is acc with all RSACMs enabled, right?

File: models/accm.py (hunk 8 above)
Comment (MichalZnalezniak) [reply models/accm.py:65]: gating_network_output_index - shape ([batch_size]) (16)
gating_network_outputs  - shape ([batch_size, self.number_of_rsacm]) (16, 8)

File: enn.py [hunk 13]
```diff
@@ -33,41 +36,112 @@ def __init__(self, *args, **kwargs):
...
+        logging.debug(f"Before Training gating networks student_model: \t Accuracy {accuracy_student}")
+        optimizers_list = []
```
Comment (bartwojcik) [inline enn.py:106]: Perhaps a more appropriate such as gating_networks_parameters?

File: enn.py [hunk 14]
```diff
@@ -33,41 +36,112 @@ def __init__(self, *args, **kwargs):
...
+            get_module_by_name(self.model, module_name).initialize_gating_network(self.teacher_model.inputs[index][0], self.args)
+            get_module_by_name(self.model, module_name).args.train_gating_networks = True
+            optimizers_list += list(get_module_by_name(self.model, module_name).gating_network.parameters())
+        gating_networks_optimizer = torch.optim.Adam(optimizers_list)
 
+        torch.autograd.set_detect_anomaly(True)
+        save_config_file(self.writer.log_dir, self.args)
+        for i, (images, labels) in enumerate(tqdm(train_loader)):
+            if i == 300:
+                break
```
Comment (bartwojcik) [inline enn.py:118]: Debugging leftover?

File: enn.py (hunk 10 above)
Comment (MichalZnalezniak) [reply enn.py:43]: By default the ACCM uses all the RSACMs (8). This codes goes sequentially ACCM by ACCM. After it goes to another ACCM you can say that it resets the number of rsacms because the previous ACCM has it rsacm set to 8.

File: enn.py (hunk 14 above)
Comment (MichalZnalezniak) [reply enn.py:118]: Yea, generally training gating_networks takes ~35minutes to iterate over train_loader. I limited training to 300 examples to check the early results.

File: enn.py (hunk 12 above)
Comment (MichalZnalezniak) [reply enn.py:104]: Yes.

File: run_experiments.py (hunk 1 above)
Comment (bartwojcik) [reply run_experiments.py:10]: I see no reason why adding gating networks should increase the amount of needed memory by that much. That's suspicious, gating networks are very small, so perhaps we somehow have a reference to tensors from previous iterations?

File: models/accm.py (hunk 8 above)
Comment (bartwojcik) [reply models/accm.py:65]: I mean the shape of features, outputs of RSACM. Let's say that the outputs of RSACMs is of shape (batch_size, number_of_rsacms, *features_size), for example (64, 8, 50, 16, 16). We want to multiply these by gating_networks_output with shape (batch_size, number_of_rsacms). For me einsum seems to be the simplest solution, something like:
`result = torch.einsum('nrchw,nr->nchw', rsacm_outputs, gating_network_outputs)`
Not sure if this is correct, I wrote it on the go. Can you test it?

File: run_experiments.py (hunk 1 above)
Comment (XYZ321COD) [reply run_experiments.py:10]: I was able to set batch_size back to `256`, after setting the 'cudnn.benchmark' to False. https://discuss.pytorch.org/t/what-does-torch-backends-cudnn-benchmark-do/5936

File: models/accm.py (hunk 8 above)
Comment (XYZ321COD) [reply models/accm.py:65]: It worked. Thanks!

File: run_experiments.py (hunk 1 above)
Comment (bartwojcik) [reply run_experiments.py:10]: Interesting. If I understood this thread correctly, the people there say that cuda kernels are cached if inputs sizes vary. Shouldn't we have a constant image size in each module? This thread: https://discuss.pytorch.org/t/set-torch-backends-cudnn-benchmark-true-consumes-huge-amount-of-memory/131010 suggests that cudnn.benchmark=True causes pytorch to choose an algorithm with huge memory requirements, which is more convincing explanation. We should remember this, it may be important later on.
